File: backend/src/Services/HumeToolSyncService.php
<?php

declare(strict_types=1);

namespace Quantis\AIPortfolioAssistant\Services;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use PDO;

/**
 * Synchronizes backend functions with Hume EVI Tools API
 *
 * This service automatically creates, updates, and manages tools in Hume EVI
 * based on the functions registered in ToolsManager.
 */
class HumeToolSyncService
{
    private Client $httpClient;
    private string $apiKey;
    private string $baseUrl;
    private ?PDO $pdo;
    private array $config;

    public function __construct(array $config, ?PDO $pdo = null)
    {
        $this->config = $config;
        $this->apiKey = $config['hume_evi']['api_key'] ?? '';
        $this->baseUrl = $config['hume_evi']['base_url'] ?? 'https://api.hume.ai/v0/evi';
        $this->pdo = $pdo;

        if (empty($this->apiKey)) {
            throw new \RuntimeException('Hume API key not configured');
        }

        $this->httpClient = new Client([
            'timeout' => 30,
            'headers' => [
                'X-Hume-Api-Key' => $this->apiKey,
                'Content-Type' => 'application/json',
            ]
        ]);
    }

    /**
     * Sync all backend tools to Hume EVI
     *
     * @param ToolsManager $toolsManager Backend tools manager
     * @return array Sync results with statistics
     */
    public function syncAllTools(ToolsManager $toolsManager): array
    {
        $localTools = $toolsManager->getToolDefinitions();
        $humeTools = $this->listHumeTools();

        $results = [
            'total_local' => count($localTools),
            'total_hume' => count($humeTools),
            'created' => [],
            'updated' => [],
            'unchanged' => [],
            'errors' => [],
            'tool_mapping' => []
        ];

        // Create mapping of Hume tools by name for quick lookup
        $humeToolsMap = [];
        foreach ($humeTools as $tool) {
            $humeToolsMap[$tool['name']] = $tool;
        }

        // Process each local tool
        foreach ($localTools as $localTool) {
            try {
                $toolName = $localTool['name'];

                if (isset($humeToolsMap[$toolName])) {
                    // Tool exists - check if update needed
                    $humeTool = $humeToolsMap[$toolName];

                    if ($this->toolNeedsUpdate($localTool, $humeTool)) {
                        $updated = $this->updateTool($humeTool['id'], $localTool);
                        $results['updated'][] = $toolName;
                        $results['tool_mapping'][$toolName] = $updated['tool_id'] ?? $updated['id'];
                    } else {
                        $results['unchanged'][] = $toolName;
                        $results['tool_mapping'][$toolName] = $humeTool['id'];
                    }
                } else {
                    // Tool doesn't exist - create it
                    $created = $this->createTool($localTool);
                    $results['created'][] = $toolName;
                    $results['tool_mapping'][$toolName] = $created['tool_id'] ?? $created['id'];
                }

                // Store mapping in database if available
                if ($this->pdo && isset($results['tool_mapping'][$toolName])) {
                    $this->storeToolMapping(
                        $toolName,
                        $results['tool_mapping'][$toolName],
                        $localTool
                    );
                }

            } catch (\Throwable $e) {
                $results['errors'][] = [
                    'tool' => $localTool['name'] ?? 'unknown',
                    'error' => $e->getMessage()
                ];
            }
        }

        // Optionally cleanup orphaned tools
        if ($this->config['hume_evi']['cleanup_orphaned_tools'] ?? false) {
            $results['deleted'] = $this->cleanupOrphanedTools($localTools, $humeTools);
        }

        return $results;
    }

    /**
     * Get synchronization status (compare local vs Hume)
     */
    public function getSyncStatus(ToolsManager $toolsManager): array
    {
        $localTools = $toolsManager->getToolDefinitions();
        $humeTools = $this->listHumeTools();

        $localNames = array_column($localTools, 'name');
        $humeNames = array_column($humeTools, 'name');

        return [
            'local_count' => count($localTools),
            'hume_count' => count($humeTools),
            'missing_in_hume' => array_values(array_diff($localNames, $humeNames)),
            'missing_in_local' => array_values(array_diff($humeNames, $localNames)),
            'needs_sync' => !empty(array_diff($localNames, $humeNames)),
            'local_tools' => $localNames,
            'hume_tools' => $humeNames
        ];
    }

    /**
     * List all tools from Hume API
     */
    public function listHumeTools(): array
    {
        try {
            $response = $this->httpClient->get("{$this->baseUrl}/tools", [
                'query' => [
                    'page_size' => 100,
                    'restrict_to_most_recent' => true
                ]
            ]);

            $data = json_decode($response->getBody()->getContents(), true);
            return $data['tools_page'] ?? [];

        } catch (GuzzleException $e) {
            throw new \RuntimeException("Failed to list Hume tools: " . $e->getMessage());
        }
    }

    /**
     * Create a new tool in Hume
     */
    public function createTool(array $toolDefinition): array
    {
        try {
            // Remove 'default' fields from schema (Hume doesn't support them)
            $cleanedSchema = $this->removeDefaultFields($toolDefinition['input_schema']);

            $payload = [
                'name' => $toolDefinition['name'],
                'parameters' => json_encode($cleanedSchema),
                'description' => $toolDefinition['description'] ?? '',
                'version_description' => 'Auto-synced from backend'
            ];

            $response = $this->httpClient->post("{$this->baseUrl}/tools", [
                'json' => $payload
            ]);

            return json_decode($response->getBody()->getContents(), true);

        } catch (GuzzleException $e) {
            throw new \RuntimeException("Failed to create tool '{$toolDefinition['name']}': " . $e->getMessage());
        }
    }

    /**
     * Remove 'default' fields from schema (Hume doesn't support them)
     */
    private function removeDefaultFields(array $schema): array
    {
        foreach ($schema as $key => &$value) {
            if ($key === 'default') {
                unset($schema[$key]);
            } elseif (is_array($value)) {
                $value = $this->removeDefaultFields($value);
            }
        }
        return $schema;
    }

    /**
     * Update an existing tool in Hume
     * Note: Hume creates new versions, doesn't update in place
     */
    public function updateTool(string $toolId, array $toolDefinition): array
    {
        // Hume uses versioning - creating new version updates the tool
        return $this->createTool($toolDefinition);
    }

    /**
     * Delete a tool from Hume
     */
    public function deleteTool(string $toolId): bool
    {
        try {
            $this->httpClient->delete("{$this->baseUrl}/tools/{$toolId}");
            return true;
        } catch (GuzzleException $e) {
            throw new \RuntimeException("Failed to delete tool: " . $e->getMessage());
        }
    }

    /**
     * Link tools to an EVI configuration
     */
    public function linkToolsToConfig(string $configId, array $toolIds): array
    {
        try {
            $tools = array_map(function($toolId) {
                return ['id' => $toolId];
            }, $toolIds);

            $response = $this->httpClient->post("{$this->baseUrl}/configs/{$configId}", [
                'json' => [
                    'tools' => $tools
                ]
            ]);

            return json_decode($response->getBody()->getContents(), true);

        } catch (GuzzleException $e) {
            throw new \RuntimeException("Failed to link tools to config: " . $e->getMessage());
        }
    }

    /**
     * Check if tool needs update
     */
    private function toolNeedsUpdate(array $localTool, array $humeTool): bool
    {
        // Clean local params first (remove default fields before comparing)
        $cleanedLocalSchema = $this->removeDefaultFields($localTool['input_schema']);
        $localParams = json_encode($cleanedLocalSchema);
        $humeParams = $humeTool['parameters'] ?? '{}';

        if ($localParams !== $humeParams) {
            return true;
        }

        // Compare description
        $localDesc = $localTool['description'] ?? '';
        $humeDesc = $humeTool['description'] ?? '';

        return $localDesc !== $humeDesc;
    }

    /**
     * Cleanup tools that exist in Hume but not in backend
     */
    private function cleanupOrphanedTools(array $localTools, array $humeTools): array
    {
        $localNames = array_column($localTools, 'name');
        $deleted = [];

        foreach ($humeTools as $humeTool) {
            if (!in_array($humeTool['name'], $localNames)) {
                try {
                    $this->deleteTool($humeTool['tool_id']);
                    $deleted[] = $humeTool['name'];
                } catch (\Throwable $e) {
                    // Log error but continue
                    error_log("Failed to delete orphaned tool {$humeTool['name']}: " . $e->getMessage());
                }
            }
        }

        return $deleted;
    }

    /**
     * Store tool mapping in database
     */
    private function storeToolMapping(string $toolName, string $humeToolId, array $definition): void
    {
        if (!$this->pdo) {
            return;
        }

        try {
            $definitionHash = md5(json_encode($definition));

            $stmt = $this->pdo->prepare(
                "INSERT INTO hume_tool_mapping (tool_name, hume_tool_id, definition_hash, last_synced)
                 VALUES (?, ?, ?, NOW())
                 ON DUPLICATE KEY UPDATE
                    hume_tool_id = VALUES(hume_tool_id),
                    definition_hash = VALUES(definition_hash),
                    last_synced = NOW()"
            );

            $stmt->execute([$toolName, $humeToolId, $definitionHash]);

        } catch (\PDOException $e) {
            // Table might not exist yet, that's okay
            error_log("Could not store tool mapping: " . $e->getMessage());
        }
    }

    /**
     * Get stored tool mappings from database
     */
    public function getStoredMappings(): array
    {
        if (!$this->pdo) {
            return [];
        }

        try {
            $stmt = $this->pdo->query("SELECT * FROM hume_tool_mapping ORDER BY tool_name");
            return $stmt->fetchAll(PDO::FETCH_ASSOC);
        } catch (\PDOException $e) {
            return [];
        }
    }

    /**
     * Test connection to Hume API
     */
    public function testConnection(): array
    {
        try {
            $tools = $this->listHumeTools();

            return [
                'success' => true,
                'message' => 'Connected to Hume API successfully',
                'tools_count' => count($tools)
            ];
        } catch (\Throwable $e) {
            return [
                'success' => false,
                'message' => 'Failed to connect: ' . $e->getMessage()
            ];
        }
    }
}

File: backend_python/app/services/hume_tool_sync_service.py
"""Port of Services/HumeToolSyncService.php.

Extracted from app/controllers/hume_tool_controller.py (Phase 7 final-review
wave, B6) — previously ported as that module's private `_HumeToolSyncService`
class with a note that no `app/services/*` module existed for it yet in this
task's file-staging guard. `hume_tool_controller.py` re-imports `HumeToolSyncService`
under its old module-private name (`_HumeToolSyncService`) so existing call
sites and tests keep working unchanged.

`guzzleErrorMessage` (a free function, no PHP class of its own) moves here
too — it exists only to format the same GuzzleHttp\\Exception\\RequestException
message this service's HTTP calls raise, and the controller's own error
handling (HumeToolController.executeTool) borrows it for the same purpose.
"""
from __future__ import annotations

import hashlib

import httpx

from app.providers._http import SHARED_SSL_CONTEXT, guzzle_body_summary
from app.services.tools_manager import ToolsManager
from app.support.logger import error_log
from app.support.phpcompat import php_empty
from app.support.phpjson import php_json_decode, php_json_encode


def guzzleErrorMessage(request: httpx.Request, response: httpx.Response) -> str:
    """Port of GuzzleHttp\\Exception\\RequestException::create()'s message format
    (Guzzle's default, no http_errors=false anywhere in HumeToolSyncService.php /
    HumeToolController.php, so every non-2xx response reaches Guzzle's own
    exception message that `$e->getMessage()` returns)."""
    status = response.status_code
    level = status // 100
    if level == 4:
        label = 'Client error'
    elif level == 5:
        label = 'Server error'
    else:
        label = 'Unsuccessful request'
    message = f"{label}: `{request.method} {request.url}` resulted in a `{status} {response.reason_phrase}` response"
    summary = guzzle_body_summary(response.text)
    if summary:
        message += f":\n{summary}\n"
    return message


class HumeToolSyncService:
    """Port of Services/HumeToolSyncService.php."""

    def __init__(self, config: dict, db=None):
        self.config = config
        humeCfg = config.get('hume_evi') if isinstance(config.get('hume_evi'), dict) else {}
        self.apiKey = humeCfg.get('api_key') if humeCfg.get('api_key') is not None else ''
        self.baseUrl = humeCfg.get('base_url') if humeCfg.get('base_url') is not None else 'https://api.hume.ai/v0/evi'
        self.db = db

        if php_empty(self.apiKey):
            raise RuntimeError('Hume API key not configured')

        self.httpClient = httpx.Client(
            timeout=30,
            verify=SHARED_SSL_CONTEXT,
            headers={'X-Hume-Api-Key': self.apiKey, 'Content-Type': 'application/json'},
        )

    def close(self) -> None:
        """Python-only addition: PHP has no equivalent — Guzzle clients die
        with the request."""
        self.httpClient.close()

    def syncAllTools(self, toolsManager: ToolsManager) -> dict:
        localTools = toolsManager.getToolDefinitions()
        humeTools = self.listHumeTools()

        results = {
            'total_local': len(localTools),
            'total_hume': len(humeTools),
            'created': [],
            'updated': [],
            'unchanged': [],
            'errors': [],
            'tool_mapping': {},
        }

        humeToolsMap = {t.get('name'): t for t in humeTools if isinstance(t, dict)}

        for localTool in localTools:
            toolName = localTool.get('name', 'unknown')
            try:
                if toolName in humeToolsMap:
                    humeTool = humeToolsMap[toolName]
                    if self._toolNeedsUpdate(localTool, humeTool):
                        updated = self.updateTool(humeTool.get('id'), localTool)
                        results['updated'].append(toolName)
                        results['tool_mapping'][toolName] = self._toolIdFrom(updated)
                    else:
                        results['unchanged'].append(toolName)
                        results['tool_mapping'][toolName] = humeTool.get('id')
                else:
                    created = self.createTool(localTool)
                    results['created'].append(toolName)
                    results['tool_mapping'][toolName] = self._toolIdFrom(created)

                if self.db is not None and results['tool_mapping'].get(toolName) is not None:
                    self._storeToolMapping(toolName, results['tool_mapping'][toolName], localTool)

            except Exception as e:  # noqa: BLE001
                results['errors'].append({'tool': toolName, 'error': str(e)})

        humeCfg = self.config.get('hume_evi') if isinstance(self.config.get('hume_evi'), dict) else {}
        cleanupOrphaned = humeCfg.get('cleanup_orphaned_tools') if humeCfg.get('cleanup_orphaned_tools') is not None else False
        if cleanupOrphaned:
            results['deleted'] = self._cleanupOrphanedTools(localTools, humeTools)

        return results

    @staticmethod
    def _toolIdFrom(data) -> object:
        """`$data['tool_id'] ?? $data['id']` — null-safe against a non-array
        decode (PHP8 warns but yields null for both, same as here)."""
        if not isinstance(data, dict):
            return None
        return data.get('tool_id') if data.get('tool_id') is not None else data.get('id')

    def getSyncStatus(self, toolsManager: ToolsManager) -> dict:
        localTools = toolsManager.getToolDefinitions()
        humeTools = self.listHumeTools()

        localNames = [t.get('name') for t in localTools]
        humeNames = [t.get('name') for t in humeTools if isinstance(t, dict)]

        missingInHume = [n for n in localNames if n not in humeNames]
        missingInLocal = [n for n in humeNames if n not in localNames]

        return {
            'local_count': len(localTools),
            'hume_count': len(humeTools),
            'missing_in_hume': missingInHume,
            'missing_in_local': missingInLocal,
            'needs_sync': len(missingInHume) > 0,
            'local_tools': localNames,
            'hume_tools': humeNames,
        }

    def listHumeTools(self) -> list:
        try:
            response = self.httpClient.get(
                f"{self.baseUrl}/tools",
                params={'page_size': 100, 'restrict_to_most_recent': 'true'},
            )
            response.raise_for_status()
            data = php_json_decode(response.text)
            return data.get('tools_page') if isinstance(data, dict) and data.get('tools_page') is not None else []
        except httpx.HTTPStatusError as e:
            raise RuntimeError('Failed to list Hume tools: ' + guzzleErrorMessage(e.request, e.response)) from e
        except httpx.RequestError as e:
            raise RuntimeError('Failed to list Hume tools: ' + str(e)) from e

    def createTool(self, toolDefinition: dict) -> dict | None:
        try:
            cleanedSchema = self._removeDefaultFields(toolDefinition.get('input_schema') or {})
            payload = {
                'name': toolDefinition.get('name'),
                'parameters': php_json_encode(cleanedSchema),
                'description': toolDefinition.get('description') if toolDefinition.get('description') is not None else '',
                'version_description': 'Auto-synced from backend',
            }
            response = self.httpClient.post(f"{self.baseUrl}/tools", json=payload)
            response.raise_for_status()
            return php_json_decode(response.text)
        except httpx.HTTPStatusError as e:
            raise RuntimeError(
                f"Failed to create tool '{toolDefinition.get('name')}': " + guzzleErrorMessage(e.request, e.response)
            ) from e
        except httpx.RequestError as e:
            raise RuntimeError(f"Failed to create tool '{toolDefinition.get('name')}': " + str(e)) from e

    def updateTool(self, toolId, toolDefinition: dict) -> dict | None:
        """Hume uses versioning — creating a new version updates the tool."""
        return self.createTool(toolDefinition)

    def deleteTool(self, toolId) -> bool:
        try:
            response = self.httpClient.delete(f"{self.baseUrl}/tools/{toolId}")
            response.raise_for_status()
            return True
        except httpx.HTTPStatusError as e:
            raise RuntimeError('Failed to delete tool: ' + guzzleErrorMessage(e.request, e.response)) from e
        except httpx.RequestError as e:
            raise RuntimeError('Failed to delete tool: ' + str(e)) from e

    def linkToolsToConfig(self, configId: str, toolIds: list) -> dict | None:
        try:
            tools = [{'id': tid} for tid in toolIds]
            response = self.httpClient.post(f"{self.baseUrl}/configs/{configId}", json={'tools': tools})
            response.raise_for_status()
            return php_json_decode(response.text)
        except httpx.HTTPStatusError as e:
            raise RuntimeError('Failed to link tools to config: ' + guzzleErrorMessage(e.request, e.response)) from e
        except httpx.RequestError as e:
            raise RuntimeError('Failed to link tools to config: ' + str(e)) from e

    def _toolNeedsUpdate(self, localTool: dict, humeTool: dict) -> bool:
        cleanedLocalSchema = self._removeDefaultFields(localTool.get('input_schema') or {})
        localParams = php_json_encode(cleanedLocalSchema)
        humeParams = humeTool.get('parameters') if humeTool.get('parameters') is not None else '{}'

        if localParams != humeParams:
            return True

        localDesc = localTool.get('description') if localTool.get('description') is not None else ''
        humeDesc = humeTool.get('description') if humeTool.get('description') is not None else ''

        return localDesc != humeDesc

    def _removeDefaultFields(self, schema):
        """Remove 'default' fields from schema (Hume doesn't support them)."""
        if isinstance(schema, dict):
            result = {}
            for key, value in schema.items():
                if key == 'default':
                    continue
                result[key] = self._removeDefaultFields(value) if isinstance(value, (dict, list)) else value
            return result
        if isinstance(schema, list):
            return [self._removeDefaultFields(v) if isinstance(v, (dict, list)) else v for v in schema]
        return schema

    def _cleanupOrphanedTools(self, localTools: list, humeTools: list) -> list:
        localNames = [t.get('name') for t in localTools]
        deleted = []

        for humeTool in humeTools:
            if not isinstance(humeTool, dict):
                continue
            if humeTool.get('name') not in localNames:
                try:
                    self.deleteTool(humeTool.get('tool_id'))
                    deleted.append(humeTool.get('name'))
                except Exception as e:  # noqa: BLE001
                    error_log(f"Failed to delete orphaned tool {humeTool.get('name')}: {e}")

        return deleted

    def _storeToolMapping(self, toolName: str, humeToolId, definition: dict) -> None:
        if self.db is None:
            return

        try:
            definitionHash = hashlib.md5(php_json_encode(definition).encode('utf-8')).hexdigest()

            self.db.execute(
                "INSERT INTO hume_tool_mapping (tool_name, hume_tool_id, definition_hash, last_synced)\n"
                " VALUES (?, ?, ?, NOW())\n"
                " ON DUPLICATE KEY UPDATE\n"
                "    hume_tool_id = VALUES(hume_tool_id),\n"
                "    definition_hash = VALUES(definition_hash),\n"
                "    last_synced = NOW()",
                [toolName, humeToolId, definitionHash],
            )
        except Exception as e:  # noqa: BLE001
            # Table might not exist yet, that's okay
            error_log(f"Could not store tool mapping: {e}")

    def getStoredMappings(self) -> list:
        if self.db is None:
            return []

        try:
            return self.db.fetch_all("SELECT * FROM hume_tool_mapping ORDER BY tool_name")
        except Exception:  # noqa: BLE001
            return []

    def testConnection(self) -> dict:
        try:
            tools = self.listHumeTools()
            return {
                'success': True,
                'message': 'Connected to Hume API successfully',
                'tools_count': len(tools),
            }
        except Exception as e:  # noqa: BLE001
            return {
                'success': False,
                'message': f'Failed to connect: {e}',
            }
